src/descriptor/mod.rs

```diff
@@ -848,7 +978,153 @@ impl<Pk: MiniscriptKey> fmt::Display for Descriptor<Pk> {
             Descriptor::Sh(ref sub) => write!(f, "sh({})", sub),
             Descriptor::Wsh(ref sub) => write!(f, "wsh({})", sub),
             Descriptor::ShWsh(ref sub) => write!(f, "sh(wsh({}))", sub),
+            Descriptor::ShSortedMulti(ref smv) => write!(f, "sh({})", smv),
+            Descriptor::WshSortedMulti(ref smv) => write!(f, "wsh({})", smv),
+            Descriptor::ShWshSortedMulti(ref smv) => write!(f, "sh(wsh({}))", smv),
+        }
+    }
+}
+
+/// Contents of a "sortedmulti" descriptor
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SortedMultiVec<Pk: MiniscriptKey, Ctx: ScriptContext> {
+    /// signatures required
+    pub k: usize,
+    /// public keys inside sorted Multi
+    pub pks: Vec<Pk>,
+    /// The current ScriptContext for sortedmulti
+    pub(crate) phantom: PhantomData<Ctx>,
+}
+
+impl<Pk: MiniscriptKey, Ctx: ScriptContext> SortedMultiVec<Pk, Ctx> {
+    fn new(k: usize, pks: &[Pk]) -> Result<Self, Error> {
+        // Check the limits before creating a new SortedMultiVec
+        // For example, under p2sh context the scriptlen can only be
+        // upto 520 bytes.
+        let term: miniscript::decode::Terminal<Pk, Ctx> = Terminal::Multi(k, Vec::from(pks));
+        let ms = Miniscript::from_ast(term)?;
+
+        // This would check all the consensus rules for p2sh/p2wsh and
+        // even tapscript in future
+        Ctx::check_ms_validity(&ms)?;
+
+        Ok(Self {
+            k,
+            pks: Vec::from(pks),
+            phantom: PhantomData,
+        })
+    }
+    /// Parse an expression tree into a SortedMultiVec
+    fn from_tree(tree: &Tree) -> Result<Self, Error>
+    where
+        <Pk as FromStr>::Err: ToString,
+    {
+        let k = tree.args[0].name.parse::<usize>().unwrap(); // FIXME: unwrap
```

Comment:
You can use `expression::parse_num` rather than `usize::parse` which will return a compatible error type so you can just use `?`.

Reply:
Done